Initial support for unity builds - speeds up compilation by up to 58%

onqtam opened the pull request:
CMake 3.16 added support for unity builds and this PR introduces them to redpanda. For a detailed description of unity builds along with the pros/cons check out [this article](https://onqtam.com/programming/2019-12-20-pch-unity-cmake-3-16/).

The measurements were done in the following way:
- Ubuntu 19.10, GCC 9.2.1, CMake 3.19, i7-8750H CPU @ 2.20GHz × 12, 32GB RAM, 1TB SSD
- 2 scenarios were tested - with up to 10 sources in a unity batch (so a target with 26 sources will end up with 3 unity files with 10/10/6 accordingly) and bundling all sources of a target into a single unity file
- measurements were taken with `time ninja` using the `real` time - CMake configuration and generation time is not included - only rebuilding whatever is left after a full build followed a clean (`ninja -t clean`)
- the first row is for a full clean build (after `ninja -t clean`) and all other rows are for minimal rebuilds by touching the respective file

|                     | baseline  | unity x10 | unity x10<br>gain % | unity x10<br>gain sec | unity all | unity all<br>gain % | unity all<br>gain sec |
|---------------------|-----------|-----------|------------------------|--------------------------|------------|-------------------------|---------------------------|
| **FULL BUILD**      | 12m5,620s | 5m57,100s | +51%                   | +368 sec                 | 5m17,877s  | +56%                    | +407 sec                  |
| `record.h`          | 10m3,643s | 4m46,272s | +52%                   | +317 sec                 | 4m15,612s  | +58%                    | +348 sec                  |
| `segment.h`         | 7m23,491s | 4m1,521s  | +45%                   | +202 sec                 | 3m25,835s  | +53%                    | +237 sec                  |
| `consensus.h`       | 5m44,449s | 2m40,766s | +53%                   | +184 sec                 | 2m7,460s   | +63%                    | +217 sec                  |
| `partition.h`       | 3m49,769s | 1m54,835s | +50%                   | +115 sec                 | 1m33,474s  | +60%                    | +137 sec                  |
| `group_manager.h`   | 3m47,872s | 2m5,896s  | +45%                   | +102 sec                 | 1m47,395s  | +53%                    | +121 sec                  |
| `kafka/client.h`    | 1m51,571s | 1m0,008s  | +46%                   | +51 sec                  | 1m8,101s   | +39%                    | +43 sec                   |
| `application.cc`    | 0m48,131s | 0m43,289s | +11%                   | +5 sec                   | 0m45,653s  | +5%                     | +2 sec                    |
| `segment.cc`        | 0m41,009s | 0m39,743s | +3%                    | +1 sec                   | 0m40,141s  | +3%                     | +1 sec                    |
| `consensus.cc`      | 0m40,537s | 0m39,641s | +2%                    | +1 sec                   | 0m39,795s  | +2%                     | +1 sec                    |
| `raft/types.cc`     | 0m34,009s | 0m40,394s | -17%                   | -6 sec                   | 0m47,479s  | -38%                    | -13 sec                   |
| `segment_reader.cc` | 0m28,268s | 0m39,084s | -39%                   | -11 sec                  | 0m59,345s  | -110%                   | -31 sec                   |
| `fetch_request.cc`  | 0m24,892s | 0m29,073s | -16%                   | -4 sec                   | 0m46,349s  | -84%                    | -21 sec                   |
| `fetch_test.cc`     | 0m24,540s | 0m39,571s | -62%                   | -15 sec                  | 0m45,652s  | -87%                    | -21 sec                   |
| `cluster_utils.cc`  | 0m19,037s | 0m32,086s | -68%                   | -13 sec                  | 0m56,767s  | -194%                   | -37 sec                   |

As can be seen in the table touching single source files might be slower in a unity build and the negative effect is more pronounced in bigger targets when all sources are clumped together. In this PR source files are batched together in groups of 10 as the better compromise but "*better*" is arguable because usually along with changes to a source file come some changes in a header file as well so perhaps batching everything in a single unity file could be worth it.

## Further work:
- stop relying on the installed version of CMake on the host system and bundle CMake 3.17.2 (or newer) in the build
- merge most test binaries into fewer ones so that unity builds can be useful (especially for storage/utils/kafka/model/etc. which have a bunch of tests in a single `CMakeLists.txt`) - there can be separate CMake commands to register separate ctests if necessary - this is a big opportunity to further cut down build times by at least 1 minute

## Other notes
- I had to install a newer version of hwloc and also copy `xxhash.c` right next to `/usr/include/xxhash.h` in order to build the project - haven't investigated in depth.
- The unity build does not consume more RAM than normal builds (or perhaps a slight increase of about 10%)
- Could produce faster binaries because unity builds are effectively the same as `-flto` (building the third-party libs as unity might be beneficial as well)
- This patch should play well with `ccache` but I haven't tested it.
- I tried using precompiled headers for the cluster/raft/kafka/storage targets and experimented with both `model/record.h` which ends up almost everywhere in `src/v/` (it produced a PCH of 300MB+ for each target) and just `<seastar/core/future.hh>` (smaller - about 80MB) but in both cases, the effects were either negligible or even detrimental - this codebase gains 0 from saving on parsing because it's dominated by template instantiation
- Overall I expected better results but the targets have very few source files and dependencies are as granular as possible and thus there isn't that much to be gained from link time with smaller object files. This codebase is of exceptional quality.

## Checklist
- [ ] Reference related [issue](https://github.com/vectorizedio/redpanda/issues)
- [ ] Update [PendingReleaseNotes.md](https://github.com/dotnwat/redpanda/blob/dev/PendingReleaseNotes.md), if relevant

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vectorizedio/redpanda?pullRequest=91) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vectorizedio/redpanda?pullRequest=91) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vectorizedio/redpanda?pullRequest=91) it.</sub>

onqtam commented on src/v/kafka/tests/group_test.cc:
```diff
@@ -47,11 +47,12 @@ static group get() {
 }
 
-static const std::vector<member_protocol> test_protos = {
+static const std::vector<member_protocol> test_group_protos = {
```
clashes with static globals in `member_test.cc` with the same names - could alternatively just exclude one of the 2 files with `SKIP_UNITY_BUILD_INCLUSION`

onqtam commented on cmake/main.cmake:
```diff
@@ -41,6 +43,18 @@ if ("${CMAKE_BUILD_TYPE}" MATCHES "Debug")
+  list(APPEND BASE_CXX_FLAGS_LIST -fcolor-diagnostics)
+  list(APPEND BASE_C_FLAGS_LIST -fcolor-diagnostics)
+endif ()
+
```
I can extract this in a separate PR if you don't want it mixed with the rest. Also, there could be an option to disable it in case of some odd circumstances.

BenPope replied:
We prefer each commit to be a minimal standalone unit of changes, each commit building on the previous ones.

Perhaps:
- fixup name collisons
- make compilation units standalone
- unity build support
- coloured output as a separate PR.

Really appreciate the effort you've put into this.

onqtam replied:
I'll bring this PR into an acceptable form in the next couple of days - will split the coloured output change into a separate PR and will force-push into this review a new set of fine-grained commits.

onqtam commented on cmake/main.cmake:
```diff
@@ -12,6 +12,8 @@ set(REDPANDA_DEPS_INSTALL_DIR ${CMAKE_CURRENT_BINARY_DIR}/deps_install
 set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_UNITY_BUILD ON)
+set(CMAKE_UNITY_BUILD_BATCH_SIZE 10)
```
This should be override-able with CMAKE_UNITY_BUILD when calling CMake. Perhaps `OFF` is a better default...

BenPope replied:
That's a great idea. Is it also possible to guard it on the CMAKE version without bumping the minimum required?

BenPope replied:
Do those flags need to be `CACHE`d?

onqtam commented on CMakeLists.txt:
```diff
@@ -1,4 +1,5 @@
-cmake_minimum_required(VERSION 3.12.0)
+# CMake 3.16 introduced support for unity builds but 3.17.2 is more stable
+cmake_minimum_required(VERSION 3.17.2)
```
This change requires that CMake 3.17.2 (or newer) is bundled with the build instead of relying on the host system

BenPope replied:
cmake has an ubuntu apt repo, so that's easy for CI. With a cmake version guard, best of both worlds.

onqtam commented:
trying with CMake 3.16 instead of 3.17.2 as the minimum required to see if the builds pass

BenPope commented:
Hi, thanks for this. I have previously had a quick play with unity builds. It seemed to be not very effective when `ccache` is in use.  

After this I suspect that link time starts to dominate, are you using `ld.gold`?

I also looked briefly at PCH, but it appears that it generates too often, especially around tests.

`iobuf.h` is the worst offender for compile time; it includes a lot and ends up everywhere. I've used https://github.com/aras-p/ClangBuildAnalyzer

onqtam commented:
@BenPope not using gold - just ld - I could play with gold as well. About PCH - even if it gets recompiled only when necessary it still delivers 0 value for this codebase.

BenPope commented:
Clang 11 has gained `-fpch-instantiate-templates` but I haven't tried it.

onqtam commented:
> Clang 11 has gained `-fpch-instantiate-templates` but I haven't tried it.

Cool - didn't know about that :)

Regarding ccache and unity builds - do you remember what the problems were? ccache should be useful for full builds with minimal changes but when impactful header files were touched (such as `iobuf.h` or anything dragged by it) then unity builds would provide compounding value.

BenPope commented:
I didn't play with unity builds extensively, at the time we only had about 55% ccache hit rates due to a compile random number generation, which I've removed. ccache should have 100% hit rate now. I don't recall any issues with unity builds and ccache, but I could imagine if they don't chunk up the files consistently, ccache would miss.

`ccache` and `ld.gold` are highly recommended.

onqtam commented:
Are there sporadic failures in ctest? 1 failed on the build: https://github.com/vectorizedio/redpanda/pull/91/checks?check_run_id=1403296370#step:6:17039
```
The following tests FAILED:
	 37 - read_materialized_topic_unit_tests_rpunit (Failed)
```

BenPope commented:
Yes, that code is under heavy development. It should be fixed soon.

senior7515 commented:
> Are there sporadic failures in ctest? 1 failed on the build: https://github.com/vectorizedio/redpanda/pull/91/checks?check_run_id=1403296370#step:6:17039
> 
> ```
> The following tests FAILED:
> 	 37 - read_materialized_topic_unit_tests_rpunit (Failed)
> ```

@onqtam thanks for this patch. @mmaslankaprv has a fix for it here https://github.com/vectorizedio/redpanda/pull/89

Waiting for v2 review. Should be fixed tom.

dotnwat commented:
Thanks for submitting this it is really cool. Especially if this speeds up CI (full builds) this will be great. As ben mentioned, optimizing for developer experience is also really important, and for that ccache is critical (as well as our never ending process of minimizing header dependences). Perhaps we could even turn on unity builds in CI if ccache issues popped up for normal development experience?

BenPope commented:
This does make a lot of sense for CI and one-time builds. My concern is introducing duplicate symbols and breaking CI without it breaking locally, but that's likely a bug anyway which we don't spot at all, and there doesn't seem to be too many at the moment. It may take a little while to settle, as the set of chunks moves around. Perhaps we can find a static analysis tool for spotting duplicate symbols ahead of time.

dotnwat commented:
good point @BenPope, sounds like we simply need an easy mechanism for turning this on and off. this PR seems to show that after all this time we've only managed to introduce a couple duplicate symbols. as long as we can reproduce locally, it seems like duplicate symbols are a pretty low risk issue.

onqtam commented:
I reworked the PR - here are the changes:
- No longer requiring a specific CMake version for unity builds - turns out all unity-related CMake code is just setting properties or global `CMAKE_*` variables and if there is no unity build support (say... version 3.15) then it will simply not be built as unity. Thus there was no need to surround any CMake code with checks. I tested this with CMake 3.15.
- Unity builds are disabled by default but now they can be enabled by `./build.sh -DCMAKE_UNITY_BUILD=ON` - that would get cached so there's no need to pass it every time - just when you are switching between `ON` and `OFF`. They are enabled on the GitHub Actions workflow.
- The coloured output has been factored into a separate PR: #114
- I decided that this still makes sense to be 1 single commit.

I just saw your concern from a few days ago about memory usage and gcc reaching up to 3gb for a single translation unit - I looked at the overall ram usage for a full build with 12 threads with ninja and concluded that it was almost the same as the non-unity build (or within 10% margin).

I also tested this with ccache and the 2 work together properly.

My opinion is that it makes even more sense to use unity builds on local machines during development and ccache on the CI (by somehow reusing the cache between builds as artifacts). My understanding is that ccache caches full translation units after the preprocessor and maps those to object files. So if I'm working locally and I pull changes which end up affectiong say... `record.h` then I would still need to rebuild almost the entire codebase (ninja will rebuild all object files where that header ends up by transitive inclusion) - ccache doesn't help with that but a unity build does. I tried `touch`-ing that header and then ccache works proeprly because a touch is not a modification ==> same contents after the preprocessor, but if I added a single empty line then almost everything got rebuilt. It makes sense to me that ccache does not help with minimal rebuilds, and in a codebase with proper dependencies minimal rebuilds are the only thing I would be doing - even when switching branches.

In my previous company we enabled unity builds everywhere - both for CI and for local workflows - the only downside to that is that you might forget to `#include` a header in some `.cc` file because it gets dragged from another `.cc` above in the unity file, and for that we had a separate CI job which built the codebase without unity, and thus we had to occasionally add a header or two.

Anyway, I don't mean to push too hard for this and perhaps I might be misunderstanding something, but you should give it a try locally with `./build.sh -DCMAKE_UNITY_BUILD=ON`.

Let me know if this is good enough.

BenPope reviewed (changes requested):
This is great.

I've played around with it and found no detriment  with `ccache`, and clearly the CI build succeeded much faster.

There are a couple of concerns:

1) It sort of breaks `compile_commands.json`, but that's not a problem for CI, and `clangd` seems to be functioning ok locally. I'm raising this in case anybody else can think of an issue it might cause, but for me, so far, it's not a show stopper.

2) The code changes are great, but for future archaeologists, it would be nicer if they were in a separate commit with an explanatory message that precedes the build system changes.

onqtam commented:
@BenPope Just force-pushed the changes into 3 separate commits.

Regarding `compile_commands.json` - there's a [bug](https://gitlab.kitware.com/cmake/cmake/-/issues/19826) reported in the CMake issue tracker. I don't think this would be too hard to fix on the CMake side and am wondering if I should take a stab at it. There is also one 1 ugly workaround for that: run CMake twice - once with `CMAKE_EXPORT_COMPILE_COMMANDS=ON` and `CMAKE_UNITY_BUILD=OFF` and then once more with `CMAKE_EXPORT_COMPILE_COMMANDS=OFF` and `CMAKE_UNITY_BUILD=ON`, but that could open another can of worms.

BenPope reviewed (approved):
LGTM.

dotnwat commented on src/v/storage/log_replayer.cc:
```diff
@@ -16,20 +16,13 @@
 #include <type_traits>
 
 namespace storage {
-static inline void crc_extend_iobuf(crc32& crc, const iobuf& buf) {
```
unused code?

onqtam replied:
no - it was identical to what's present in `bytes/utils.h` and I included it instead of this snippet

dotnwat commented on src/v/cluster/tests/commands_serialization_test.cc:
```diff
@@ -29,13 +29,9 @@
 #include <bits/stdint-intn.h>
```
Please keep the commit subject line limited to around 50 characters.

dotnwat replied:
github doesn't support comments about commit message, so this is just a comment on the first LOC in the commit but is in refernence to the commit message.

onqtam replied:
`CONTRIBUTING.md` would need updating in that case. I'll re-force-push with shorter commit subject lines.

dotnwat replied:
Thanks, I'll get that file updated.

dotnwat reviewed (changes requested):
Just a tiny change to the commit message, please.

onqtam commented on src/v/cluster/tests/commands_serialization_test.cc:
```diff
@@ -29,13 +29,9 @@
-using namespace std::chrono_literals;
+#include "topic_table_fixture.h"
 
-static std::unique_ptr<cluster::allocation_node>
```
it's not unused, but it's identical to what is already present in `"topic_table_fixture.h"` - would that require an explanation?

dotnwat reviewed (approved):
thanks @onqtam this is a nice step forward for the build system!
